File: src/python/core/src/dicebear/style_def/component_variant.py
"""Read-only view over an entry in a component's ``variants`` block."""

from __future__ import annotations

from typing import Any, cast

from .element import Element


class ComponentVariant:
    """Read-only view over an entry in a component's ``variants`` block."""

    def __init__(self, data: dict[str, Any]) -> None:
        self._data = data
        self._elements: list[Element] | None = None

    def elements(self) -> list[Element]:
        """Return the variant's elements, lazily wrapped on first access."""
        if self._elements is None:
            self._elements = [Element(el) for el in self._data["elements"]]

        return self._elements

    def weight(self) -> int | float:
        """Return the weighted-pick weight for this variant, defaulting to ``1``."""
        return cast("int | float", self._data.get("weight", 1))

File: src/go/core/internal/style/component_variant.go
package style

import "strings"

// ComponentVariant is a read-only view over an entry in a component's variants
// block.
type ComponentVariant struct {
	Elements []Element `json:"elements"`
	Weight   *float64  `json:"weight"`
	Tags     []string  `json:"tags"`
}

// WeightOr1 returns the weighted-pick weight, defaulting to 1.
func (cv ComponentVariant) WeightOr1() float64 {
	if cv.Weight != nil {
		return *cv.Weight
	}
	return 1
}

// HasTag tests this variant against a single tag-filter token's grammar. With an
// empty value, it matches a whole category: the bare category tag or any
// category:value tag. With a value, it matches only the exact category:value tag.
// The resolver composes these checks into the allow/disallow filter structure.
func (cv ComponentVariant) HasTag(category, value string, hasValue bool) bool {
	if !hasValue {
		prefix := category + ":"
		for _, tag := range cv.Tags {
			if tag == category || strings.HasPrefix(tag, prefix) {
				return true
			}
		}
		return false
	}

	want := category + ":" + value
	for _, tag := range cv.Tags {
		if tag == want {
			return true
		}
	}
	return false
}
